### PR TITLE
Refactor S3 store creation

### PR DESCRIPTION
**Pull request by simonbohnen**
- Move URL parsing logic of S3FSStores (replacing `Boto3Store` and `BotoStore`) to the S3FSStore
- Deprecated old store creation methods

If the changes in this PR look good, I will create separate PRs for the other stores from #71.

* [x] Added a `docs/changes.rst` entry

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -110,5 +110,10 @@ Pipfile.lock
...
 # minimalkv
 store/
+old/
+
 # Exploratory code
-exp.py
+exploration_scripts/
+
+# terraform
+terraform/.terraform*
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply .gitignore:119]: Right, I extracted the terraform stuff to #75 👌🏼

File: minimalkv/_get_store.py [hunk 2]
```diff
@@ -52,7 +60,96 @@ def get_store_from_url(url: str) -> KeyValueStore:
...
+
+def extract_wrappers(parsed_url: SplitResult) -> List[str]:
+    """
+    Extract wrappers from a parsed URL.
+
+    Wrappers allow you to add additional functionality to a store, e.g. encryption.
+    Wrappers are specified in the fragment part of the URL, e.g. "s3://...#wrap:readonly+urlencode"
+
+    Wrappers can also be specified as part of the scheme, e.g. "s3+readonly+urlencode://...".
+    This is deprecated and will be removed in a future version.
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/_get_store.py:116]: Right, wrappers in the scheme are deprecated. I'll clarify that. And no, we don't need the function to be private 👌🏼

File: minimalkv/_get_store.py [hunk 3]
```diff
@@ -118,12 +215,20 @@ def get_store(
...
+        DeprecationWarning,
+        stacklevel=2,
+    )
+    from minimalkv._old_store_creation import create_store
     from minimalkv._store_decoration import decorate_store
 
     # split off old-style wrappers, if any:
     parts = type.split("+")
-    type = parts.pop(-1)
+    type = parts.pop(0)
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/_get_store.py:231]: The store type is the first part of the scheme (index `0`). Popping off the last one here is just wrong 🤷🏼‍♂️

File: minimalkv/_key_value_store.py [hunk 4]
```diff
@@ -462,6 +464,48 @@ def __exit__(
...
+        Returns
+        -------
+        store : KeyValueStore
+            Store created from URL.
+        """
+        from minimalkv import get_store_from_url
+
+        store = get_store_from_url(url, store_cls=cls)
+        if not isinstance(store, cls):
+            raise ValueError(f"Expected {cls}, got {type(store)}")
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/_key_value_store.py:485]: I think I'm going to get rid of this method actually.

File: minimalkv/_key_value_store.py [hunk 5]
```diff
@@ -462,6 +464,48 @@ def __exit__(
...
+            The parsed URL.
+        query: Dict[str, str]
+            Query parameters from the URL.
+
+        Returns
+        -------
+        store : KeyValueStore
+            The created KeyValueStore.
+        """
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/_key_value_store.py:507]: Only `get_store_from_url` should be public 👌🏼

File: minimalkv/_key_value_store.py [hunk 6]
```diff
@@ -462,6 +464,48 @@ def __exit__(
         """
         self.close()
 
+    @classmethod
+    def from_url(cls, url: str) -> "KeyValueStore":
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/_key_value_store.py:468]: Agreed, I've removed this method.

File: minimalkv/_old_urls.py [hunk 7]
```diff
@@ -32,15 +33,28 @@ def url2dict(url: str, raise_on_extra_params: bool = False) -> Dict[str, Any]:
...
+    # We will just use the last occurrence for each key
+    query = {k: v[-1] for k, v in query_listdict.items()}
+
     parsed = dict(
         scheme=u.getscheme(),
         host=u.gethost(),
         port=u.getport(),
         path=u.getpath(),
-        query=u.getquerydict(),
+        query=query,
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/_old_urls.py]: This should be part of a different PR 👌🏼

File: minimalkv/_boto.py [hunk 8]
```diff
@@ -10,13 +10,23 @@ def _get_s3bucket(
...
+        "calling_format": OrdinaryCallingFormat(),
+    }
+
+    # Split up the host into host and port.
+    if ":" in host:
+        host, port = host.split(":")
+        s3_connection_params["host"] = host
+        s3_connection_params["port"] = int(port)
+    else:
+        s3_connection_params["host"] = host
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/_boto.py]: `urllib` only supports full URLs, i.e. requires a scheme first. If you pass e.g. `127.0.0.1:8000` in, `127.0.0.1` is recognized as the scheme. I think doing this manually should respect all the URL format rules.

File: minimalkv/_get_store.py [hunk 9]
```diff
@@ -32,27 +41,110 @@ def get_store_from_url(url: str) -> KeyValueStore:
...
+    old_wrappers = list(reversed(parts))
+
+    if old_wrappers:
+        warnings.warn(
+            """
+            Using wrappers as part of the scheme is deprecated. Please specify wrappers as part of the fragment instead,
+            e.g. "s3://...#wrap:readonly+urlencode" instead of "s3+readonly+urlencode://...".
+            """,
+            DeprecationWarning,
+        )
```
Comment (janjagusch) [inline minimalkv/_get_store.py:127]: This was already previously deprecated, right? It's nothing we've introduced in this PR, right?

File: minimalkv/_get_store.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (janjagusch) [reply minimalkv/_get_store.py:226]: I don't want to be too pedantic, but could we move this into a separate mini-PR?

File: minimalkv/net/s3fsstore.py [hunk 10]
```diff
@@ -76,3 +82,132 @@ def _url_for(self, key) -> str:
         return self._fs.url(
             f"{self.bucket.name}/{self.object_prefix}{key}", expires=self.url_valid_time
         )
+
+    def __eq__(self, other):
+        """
+        Assert that two ``S3FSStore``s are equal.
```
Comment (janjagusch) [inline minimalkv/net/s3fsstore.py:88]: I wouldn't call it assert, as there is no assert statement in the function body.

File: tests/bucket_manager.py [hunk 11]
```diff
@@ -73,35 +97,24 @@ def boto3_bucket(
...
         )
 
     bucket = s3_resource.Bucket(name)
-
-    yield bucket
-
-    for key in bucket.objects.all():
-        key.delete()
-    bucket.delete()
+    return bucket
```
Comment (janjagusch) [inline tests/bucket_manager.py:117]: We don't we clean up the bucket anymore?

**Review by janjagusch — COMMENTED** [review]
getting there!

File: minimalkv/net/boto3store.py [hunk 12]
```diff
@@ -245,3 +248,23 @@ def _url_for(self, key):
...
+            isinstance(other, Boto3Store)
+            and self.bucket.name == other.bucket.name
+            and self.bucket.meta.client.meta.endpoint_url
+            == other.bucket.meta.client.meta.endpoint_url
+            and self._object_prefix == other._object_prefix
+            and self.url_valid_time == other.url_valid_time
+            and self.reduced_redundancy == other.reduced_redundancy
+            and self.public == other.public
+            and self.metadata == other.metadata
+        )
```
[the remark this answers is not in the document]
Comment (janjagusch) [reply minimalkv/net/boto3store.py:270]: Let's move this to the tests. I'd be hesitant to implement the `__eq__` operator for only one store class. This might cause confusion as a user.

File: minimalkv/_get_store.py (hunk 9 above)
Comment (simonbohnen) [reply minimalkv/_get_store.py:127]: There was no warning given previously, so we should probably keep it like that for now.

File: tests/bucket_manager.py (hunk 11 above)
Comment (simonbohnen) [reply tests/bucket_manager.py:117]: I split the methods up into creating a bucket reference and creating the bucket itself, which also includes cleaning up the bucket. See [here](https://github.com/data-engineering-collective/minimalkv/pull/76/files#diff-2d7b489f836852c4047e7cbfb7b5f8cca1fd388d7acf20b86a965fb1feb9e299R72).

File: minimalkv/_get_store.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/_get_store.py:216]: Makes sense 👍🏼

**Review by janjagusch — APPROVED** [review]

File: environment.yml [hunk 13]
```diff
@@ -6,6 +6,7 @@ dependencies:
   - azure-storage-blob
   - boto
   - boto3
+  - mypy-boto3-s3
```
Comment (xhochy) [inline environment.yml:9]: This should be part of the pre-commit hooks here: https://github.com/data-engineering-collective/minimalkv/blob/20a516329605cbe0c776975cc1ee01a97f8c7dc0/.pre-commit-config.yaml#L29

File: minimalkv/_get_store.py [hunk 14]
```diff
@@ -32,27 +40,101 @@ def get_store_from_url(url: str) -> KeyValueStore:
...
 
-        * DictStore: ``memory://``
-        * RedisStore: ``redis://[[password@]host[:port]][/db]``
-        * FilesystemStore: ``fs://path``
-        * BotoStore ``s3://access_key:secret_key@endpoint/bucket[?create_if_missing=true]``
-        * AzureBlockBlockStorage: ``azure://account_name:account_key@container[?create_if_missing=true]``
-        * AzureBlockBlockStorage (SAS): ``azure://account_name:shared_access_signature@container?use_sas&create_if_missing=false``
-        * AzureBlockBlockStorage (SAS): ``azure://account_name:shared_access_signature@container?use_sas&create_if_missing=false[?max_connections=2&socket_timeout=(20,100)]``
-        * AzureBlockBlockStorage (SAS): ``azure://account_name:shared_access_signature@container?use_sas&create_if_missing=false[?max_connections=2&socket_timeout=(20,100)][?max_block_size=4*1024*1024&max_single_put_size=64*1024*1024]``
-        * GoogleCloudStorage: ``gcs://<base64 encoded credentials JSON>@bucket_name[?create_if_missing=true][&bucket_creation_location=EUROPE-WEST1]``
```
Comment (xhochy) [inline minimalkv/_get_store.py:43]: Why did you remove the other stores?

File: minimalkv/_get_store.py [hunk 15]
```diff
@@ -32,27 +40,101 @@ def get_store_from_url(url: str) -> KeyValueStore:
...
+    # split off old-style wrappers, if any:
+    parts = parsed_url.getscheme().split("+")
+    # pop off the type of the store
+    parts.pop(0)
+    old_wrappers = list(reversed(parts))
+
+    # find new-style wrappers, if any:
+    fragment = parsed_url.getfragment()
+    fragments = fragment.split("#") if fragment else []
+    wrap_spec = list(filter(lambda s: s.startswith("wrap:"), fragments))
```
Comment (xhochy) [inline minimalkv/_get_store.py:122]: Use the simpler version with `[…]`
```suggestion
    wrap_spec = [s for s in fragments if s.startswith("wrap:")]
```

File: minimalkv/_get_store.py [hunk 16]
```diff
@@ -32,27 +40,101 @@ def get_store_from_url(url: str) -> KeyValueStore:
...
+        The list of wrappers.
     """
-    return get_store(**url2dict(url))
+    # split off old-style wrappers, if any:
+    parts = parsed_url.getscheme().split("+")
+    # pop off the type of the store
+    parts.pop(0)
+    old_wrappers = list(reversed(parts))
+
+    # find new-style wrappers, if any:
```
Comment (xhochy) [inline minimalkv/_get_store.py:119]: Can you write a more descriptive comment on what we expect here.

File: minimalkv/_get_store.py [hunk 17]
```diff
@@ -32,27 +40,101 @@ def get_store_from_url(url: str) -> KeyValueStore:
...
+def _extract_wrappers(parsed_url: SplitResult) -> List[str]:
+    """
+    Extract wrappers from a parsed URL.
+
+    Wrappers allow you to add additional functionality to a store, e.g. encryption.
+    Wrappers are specified in the fragment part of the URL, e.g. "s3://...#wrap:readonly+urlencode"
 
-    .. code-block:: python
+    Deprecated: wrappers can also be specified as part of the scheme, e.g. "s3+readonly+urlencode://...".
+    This way of specifying wrappers will be removed in a future version.
```
Comment (xhochy) [inline minimalkv/_get_store.py:101]: Why? I prefer that syntax. It is also a similar approach as I have seen in Hadoop and SQLAlchemy.

File: minimalkv/_get_store.py [hunk 18]
```diff
@@ -118,6 +200,14 @@ def get_store(
         Key value store of type ``type`` as described in ``kwargs`` parameters.
 
     """
+    warn(
```
Comment (xhochy) [inline minimalkv/_get_store.py:203]: No, as said before, I don't want to remove that interface.

**Review by xhochy — CHANGES_REQUESTED** [review]

**Review by xhochy — APPROVED** [review]
